### PR TITLE
Bug 62870 / Templates : Add ability to provide parameters

### PR DESCRIPTION
**Pull request by ubikloadpack**
## Description

This is the implementation of a feature in enhancement list:

"Templates : Add ability to provide parameters:"

- https://bz.apache.org/bugzilla/show_bug.cgi?id=62870

It has been implemented by members of UbikLoadPack (https://ubikloadpack.com) team (Thomas, Florent and Philippe) and is donated to Apache JMeter project since we have signed a CCLA few years ago.

## Motivation and Context
This feature will allow providing a kind of wizard for template allowing to specify parameters when creating a new Plan:

https://bz.apache.org/bugzilla/show_bug.cgi?id=62870

## How Has This Been Tested?

The PR contains JUnit tests for part of the code.
And we have performed manual testing.

## Screenshots (if appropriate):

The feature is fully documented in the PR with screenshots as long as existing one (template).

screen shot 2018-11-19 at 22 34 37

## Types of changes

- New feature (non-breaking change which adds functionality)
- No Breaking change as parameters are optional

## Checklist:

- [X] My code follows the [code style][style-guide] of this project.
- [X] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

### CODE REVIEW

**Comment by pmouawad** [comment]
Hello Team,
The PR looks good to me, tested on Mac OSX. 
Anybody willing to commit it ?
I can take in charge , but I would prefer somebody else reviewing it.
If no feedback by tomorrow evening, I'll commit it.

Thanks

**Comment by FSchumacher** [comment]
This seems to be quite a big patch. Do you think two days are enough to have a good look at it?

**Comment by pmouawad** [comment]
Hi Felix,
You’re right.

Take your time 

Sorry, I am always impatient guy :)

Regards

**Review by FSchumacher — APPROVED** [review]
The feature seems great and the implementation looks good (apart from the minor nit picks, that yo will always find when looking at someone else code).

**Comment by pmouawad** [comment]
Hello Felix, 
I have committed the fixes following your review.
Let me know if you will commit the PR (which I would prefer), otherwise I'll do it.

Thanks
Regards

**Comment by pmouawad** [comment]
Hello,
I'll be merging it today.

Regards

File: src/core/org/apache/jmeter/gui/action/SelectTemplatesDialog.java [hunk 1]
```diff
@@ -137,67 +160,138 @@ private void checkDirtyAndLoad(final ActionEvent actionEvent)
...
+                log.warn("Could not delete generated output file {} from template {}", temporaryGeneratedFile, templateFile);
+            }
+        }
+    }
+
+    /**
+     * @param actionEvent {@link ActionEvent}
+     * @return true if plan is not dirty or has been saved 
+     */
+    boolean checkDirty(final ActionEvent actionEvent) {
```
Comment (FSchumacher) [inline src/core/org/apache/jmeter/gui/action/SelectTemplatesDialog.java]: Any reason for not using `private`?

File: src/core/org/apache/jmeter/gui/action/SelectTemplatesDialog.java [hunk 2]
```diff
@@ -137,67 +160,138 @@ private void checkDirtyAndLoad(final ActionEvent actionEvent)
...
+        } finally {
+            if (temporaryGeneratedFile != null && !temporaryGeneratedFile.delete()) {
+                log.warn("Could not delete generated output file {} from template {}", temporaryGeneratedFile, templateFile);
+            }
+        }
+    }
+
+    /**
+     * @param actionEvent {@link ActionEvent}
+     * @return true if plan is not dirty or has been saved
```
Comment (FSchumacher) [inline src/core/org/apache/jmeter/gui/action/SelectTemplatesDialog.java]: Somehow I think a method `checkDirty` would return `true` if plan is *dirty* and has not been saved. Can you think of a better name?

**Pull request closed without merging**